linkerd/distribute/src/lib.rs

```diff
@@ -0,0 +1,258 @@
+    Empty,
+
+    /// A distribution that uses the first available backend in an ordered list.
+    FirstAvailable(Arc<[K]>),
+
+    /// A distribution that uses the first available backend when randomly
+    /// selecting over a weighted distributino of backends.
+    RandomAvailable(Arc<WeightedKeys<K>>),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct WeightedKeys<K> {
+    keys: Vec<K>,
+    index: WeightedIndex<u32>,
+}
+
+#[derive(Debug)]
+enum Inner<K> {
+    Empty,
+    FirstAvailable {
+        keys: Arc<[K]>,
+    },
+    RandomAvailable {
+        keys: Arc<WeightedKeys<K>>,
+        polled_idxs: HashSet<usize>,
+        rng: SmallRng,
+    },
+}
+
+// === impl NewDistribute ===
+
+impl<K: Hash + Eq, S> NewDistribute<K, S> {
+    #[inline]
+    pub fn new(pairs: impl IntoIterator<Item = (K, S)>) -> Self {
+        Self {
+            backends: Arc::new(pairs.into_iter().collect()),
+        }
+    }
+}
+
+impl<T, K, S> NewService<T> for NewDistribute<K, S>
+where
+    T: Param<Distribution<K>>,
+    K: Hash + Eq + Clone,
+    S: Clone,
+{
+    type Service = Distribute<K, S>;
+
+    fn new_service(&self, target: T) -> Self::Service {
+        Self::Service {
+            inner: target.param().into(),
+            backends: self.backends.clone(),
+            ready: None,
+        }
+    }
+}
+
+// === impl Distribute ===
+
+impl<K: Clone, S: Clone> Clone for Distribute<K, S> {
+    fn clone(&self) -> Self {
+        Self {
+            backends: self.backends.clone(),
+            inner: self.inner.clone(),
+            // Clear the ready index so that the new clone must become ready
+            // independently.
+            ready: None,
+        }
+    }
+}
+
+impl<Req, K, S> Service<Req> for Distribute<K, S>
+where
+    K: Hash + Eq + std::fmt::Debug,
+    S: Service<Req> + Clone,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    // Note that this doesn't necessarily drive all inner services to readiness.
+    // We expect that these inner services should be buffered or otherwise drive
+    // themselves to readiness (i.e. via SpawnReady).
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // If we've already chosen a ready index, then skip polling.
+        if self.ready.is_some() {
+            return Poll::Ready(Ok(()));
+        }
+
+        match self.inner {
+            Inner::Empty => {
+                tracing::debug!("empty distribution will never become ready");
+            }
+
+            Inner::FirstAvailable { ref keys } => {
+                for key in keys.iter() {
+                    let svc = self
+                        .backends
+                        .get(key)
+                        .expect("distributions must not reference unknown backends");
+                    let mut svc = svc.clone();
+                    if svc.poll_ready(cx)?.is_ready() {
+                        self.ready = Some(svc);
```

Comment:
how will this interact with `Service`s where cloning clears readiness state? AFAICT, if the inner service's readiness is per-clone, and we return `Poll::Pending` because no inner service is ready, the task will not be woken by one of those inner services becoming ready, since the clones are all dropped immediately after polling them.

would it make more sense to use `get_mut` here, and poll the service instance stored in `self.backends`, rather than cloning it? we could change `self.ready` to store the _key_ [^1] of the ready service, rather than a clone of that service, and have `call` look up the service for that key.

[^1]: since the services live in an `IndexMap`, it would be a bit more efficient to use [`get_full_mut`](https://docs.rs/indexmap/latest/indexmap/map/struct.IndexMap.html#method.get_full_mut) and store the ready service's index rather than its key --- this is morally equivalent but has the advantage of not having to hash the key, and doesn't clone the key if the keys are expensive to clone (e.g. strings)...

Reply:
that's an excellent point. We can't use get_*_mut currently because the index is arc'd to make cloning cheap. I think it's probably a premature optimization, though; so it would be better to clone the indexmap. (we expect the number of backends to be fairly small).

A prior version stored `ready_idx` instead of `ready`. I agree it probably makes sense to go back to that.